Add raster-dem encoding parameters to RasterDEMSourceSpecification

ibesora opened the pull request:
Following #326 this PR adds `redMix`, `blueMix`, `greenMix` and `baseMix` parameters to `RasterDEMSourceSpecification`

See [this](https://github.com/maplibre/maplibre-gl-js/pull/3087) PR on the maplibre-gl-js repo for the other part of the implementation

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->


 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

HarelM commented:
I'm not sure the word mix is the right word here.
I would consider redMultiplier or redFactor as better describing the expected definition.
For the base I would consider baseShift maybe.

HarelM commented:
Do we want to add specific valuations (and tests) for when setting the new value with mapbox ro terrain encoding? Are there any other tests that needs to be added?

ibesora commented:
> I would consider redMultiplier or redFactor as better describing the expected definition.
For the base I would consider baseShift maybe.

Changed them to `redFactor`, `greenFactor`, `blueFactor`, and `baseShift`

> Do we want to add specific valuations (and tests) for when setting the new value with mapbox ro terrain encoding? Are there any other tests that needs to be added?

Added an error raised when "custom" encoding is set and you the style also sets any of the encoding parameters

HarelM commented on src/validate/validate_raster_dem_source.ts:
```diff
@@ -0,0 +1,53 @@
+
+    for (const key in rasterDEM) {
+        if (!isCustomEncoding && customEncodingKeys.includes(key)) {
+            errors.push(new ValidationError(key, rasterDEM[key], `"${key}" is only valid when "encoding" is set to "custom". ${options.value.encoding} encoding found`));
```
Consider adding the source name here if it's not returned - sometimes these error and not easy to find in big style files...

HarelM commented:
I've added a few nit picks, looks good otherwise! THANKS!

wipfli commented:
How did you decided that it should be `redFactor` and not `red-factor` or `red_factor`? Skimming over the json file, it looks like all 3 conventions can be found...

ibesora commented on src/validate/validate.ts:
```diff
@@ -48,6 +49,7 @@ const VALIDATORS = {
     'padding': validatePadding,
     'variableAnchorOffsetCollection': validateVariableAnchorOffsetCollection,
     'sprite': validateSprite,
+    'source_raster_dem': validateRasterDEMSource
```
You are right, I was just going off the 'terrain' validation and missed this bit

HarelM replied:
This was marked as resolved, but I still see it here, is it needed?

ibesora commented:
> I've added a few nit picks, looks good otherwise! THANKS!

Updated, thanks for reviewing!

>How did you decided that it should be redFactor and not red-factor or red_factor? Skimming over the json file, it looks like all 3 conventions can be found...

This was not clear to me. I did look at how the other properties of sources were called and everything, except `minzoom` and `maxzoom` (🤷) , seemed to use camelcase (`tileSize`, `promoteId`, `clusterRadius`, `clusterMaxZoom`, ...)

HarelM commented:
Can you check that the coverage for the newly added file is 100%?
For some reason, the codecov is not getting updated with latest run information, I'll need to check what went wrong, but until then let's do it manually...
Other then the last comment, this PR is good to go from my point of view.

HarelM commented:
Yes, after meeting this there's a need to crave a release and then use it in gl js.
The best approach would be for your to create the version bump PR here so that I can review it and publish a new version.
